src/components/Reports/GoalCalculator/SharedComponents/GoalCalculatorGrid/GoalCalculatorGrid.tsx

```diff
@@ -1,196 +1,365 @@
 import { StyledGrid } from './StyledGrid';
 
 const StyledCard = styled(Card)(({ theme }) => ({
   borderRadius: theme.shape.borderRadius,
   paddingTop: theme.spacing(2),
 }));
 
-const StyledAddButton = styled(Button)({
-  color: 'primary.main',
-});
-
-const StyledBox = styled(Box)(({ theme }) => ({
-  display: 'flex',
-  justifyContent: 'flex-start',
-  marginBottom: theme.spacing(2),
+const ErrorCell = styled(Box)(({ theme }) => ({
+  height: '100%',
+  border: `2px solid ${theme.palette.error.main}`,
 }));
 
-const StyledGridContainer = styled(Box)({
-  height: 'auto',
-  width: '100%',
-});
-
-export interface GoalCalculatorGridFormValues {
-  gridData: Array<{
-    id: number;
-    name: string;
-    amount: number;
-  }>;
-  lumpSumAmount: number;
-}
-
 interface GoalCalculatorGridProps {
-  formData?: Array<{
-    id: number;
-    name: string;
-    amount: number;
-  }>;
+  category: BudgetFamilyFragment['primaryBudgetCategories'][number];
   promptText?: string;
-  categoryName: string;
 }
 
+// Yup validation schemas
+const directInputSchema = yup.object({
+  amount: yup
+    .number()
+    .min(0, 'Amount must be positive')
+    .required('Amount is required'),
+});
+
+const subBudgetCategorySchema = yup.object({
+  label: yup
+    .string()
+    .trim()
+    .min(1, 'Label is required')
+    .max(100, 'Label must be less than 100 characters')
+    .required('Label is required'),
+  amount: yup
+    .number()
+    .min(0, 'Amount must be positive')
+    .required('Amount is required'),
+});
+
 export const GoalCalculatorGrid: React.FC<GoalCalculatorGridProps> = ({
-  formData,
+  category,
   promptText,
-  categoryName,
 }) => {
-  const { handleContinue } = useGoalCalculator();
   const { t } = useTranslation();
+  const locale = useLocale();
+  const { label: categoryName } = category;
+  const accountListId = useAccountListId() ?? '';
 
-  const initialValues: GoalCalculatorGridFormValues = {
-    gridData: formData || [
-      { id: 1, name: 'Freelance Work', amount: 2500 },
-      { id: 2, name: 'Investment Returns', amount: 1200 },
-      { id: 3, name: 'Rental Income', amount: 1800 },
-    ],
-    lumpSumAmount: 0,
-  };
+  const gridData = React.useMemo(
+    () =>
+      category.subBudgetCategories.map((subCategory) => ({
+        id: subCategory.id,
+        label: subCategory.label,
+        amount: subCategory.amount,
+        canDelete: !subCategory.category,
+      })),
+    [category.subBudgetCategories]
+  );
+
+  const totalAmount = gridData.reduce((sum, item) => sum + item.amount, 0);
+  const [cellErrors, setCellErrors] = useState<
+    Record<string, string | undefined>
+  >({});
+  const [directInputError, setDirectInputError] = useState<string>('');
+  const [inputValue, setInputValue] = useState<string>('');
+  const [updatePrimaryBudgetCategory] =
+    useUpdatePrimaryBudgetCategoryMutation();
+  const [localDirectInput, setLocalDirectInput] = useState<boolean | null>(
```

Comment:
What is this state for? Can we use the budget category's `directInput` field instead since we are using optimistic updates?

Or is it for caching the last direct input value the user entered?